pkg/agent/controller/servicecidr/controller.go

```diff
@@ -0,0 +1,125 @@
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package servicecidr
+
+import (
+	"net"
+	"sync"
+	"time"
+
+	corev1 "k8s.io/api/core/v1"
+	coreinformers "k8s.io/client-go/informers/core/v1"
+	"k8s.io/client-go/tools/cache"
+	"k8s.io/klog/v2"
+	utilnet "k8s.io/utils/net"
+
+	"antrea.io/antrea/pkg/agent/util"
+)
+
+const (
+	ipv4AddrLength = 32
+	ipv6AddrLength = 128
+
+	// Disable resyncing.
+	resyncPeriod time.Duration = 0
+)
+
+type Controller struct {
+	serviceInformer cache.SharedIndexInformer
+	sync.RWMutex
+	clusterIPv4CIDR *net.IPNet
+	clusterIPv6CIDR *net.IPNet
+}
+
+func NewServiceCIDRController(serviceInformer coreinformers.ServiceInformer) *Controller {
+	c := &Controller{
+		serviceInformer: serviceInformer.Informer(),
+	}
+
+	c.serviceInformer.AddEventHandlerWithResyncPeriod(
+		cache.ResourceEventHandlerFuncs{
+			AddFunc:    c.addService,
+			UpdateFunc: c.updateService,
+		},
+		resyncPeriod,
+	)
+	return c
+}
+
+func (c *Controller) GetClusterIPCIDR(isIPv6 bool) *net.IPNet {
+	c.RLock()
+	defer c.RUnlock()
+	if isIPv6 {
+		return c.clusterIPv6CIDR
+	} else {
+		return c.clusterIPv4CIDR
+	}
+}
+
+func (c *Controller) addService(obj interface{}) {
+	svc := obj.(*corev1.Service)
+	c.updateClusterIPCIDR(svc)
+}
+
+func (c *Controller) updateService(_, obj interface{}) {
+	svc := obj.(*corev1.Service)
+	c.updateClusterIPCIDR(svc)
+}
+
+func (c *Controller) updateClusterIPCIDR(svc *corev1.Service) {
+	clusterIPs := svc.Spec.ClusterIPs
+	if len(clusterIPs) == 0 {
+		return
+	}
+
+	c.Lock()
+	defer c.Unlock()
+	klog.V(2).InfoS("Processing Service ADD or UPDATE event", "Service", klog.KObj(svc))
+	for _, clusterIPStr := range clusterIPs {
+		clusterIP := net.ParseIP(clusterIPStr)
+		isIPv6 := utilnet.IsIPv6(clusterIP)
+
+		curClusterIPCIDR := c.clusterIPv4CIDR
+		mask := ipv4AddrLength
+		if isIPv6 {
+			curClusterIPCIDR = c.clusterIPv6CIDR
+			mask = ipv6AddrLength
+		}
+
+		if curClusterIPCIDR != nil && curClusterIPCIDR.Contains(clusterIP) {
+			continue
+		}
+
+		var newClusterIPCIDR *net.IPNet
+		var err error
+		if curClusterIPCIDR != nil {
+			// If the calculated ClusterIP CIDR exists and contains the ClusterIP, calculate a new ClusterIP CIDR by
+			// enlarging the current ClusterIP CIDR with the ClusterIP.
+			if newClusterIPCIDR, err = util.ExtendCIDRWithIP(curClusterIPCIDR, clusterIP); err != nil {
+				klog.ErrorS(err, "Error when enlarging the destination CIDR", "curClusterIPCIDR", curClusterIPCIDR, "clusterIP", clusterIP)
```

Comment:
"Error when enlarging the destination CIDR" -> "Error when enlarging the Service ClusterIP CIDR"
"curClusterIPCIDR" -> "ClusterIPCIDR"